src/client/mod.rs

```diff
@@ -290,175 +375,260 @@ impl Client {
-            if type_code == MTCode::String as u32 {
-                // value
-                c.write_u64::<Endian>(0)?;
-                // extra (string offset)
-                string_val_off_off = Some(c.position());
-                c.write_u64::<Endian>(0)?;
-            } else {
-                value_offset = c.position() as usize;
-                value_size = NUMERIC_VALUE_SIZE;
+            // write instances and record their offsets
+            let mut instance_blk_offs = Vec::with_capacity(indom.instances.len());
+            for instance in &indom.instances {
+                c.set_position(instance_blk_off);
+
+                // indom offset
+                c.write_u64::<Endian>(indom_off)?;
+                // zero pad
+                c.write_u32::<Endian>(0)?;
+                // instance id
+                c.write_u32::<Endian>(Indom::instance_id(&instance))?;
+                // instance
+                c.write_all(instance.as_bytes())?;
+                c.write_all(&[0])?;
 
-                // value
-                m.write_val(&mut c)?;
-                // extra (string offset)
-                c.write_u64::<Endian>(0)?;
+                instance_blk_offs.push(instance_blk_off);
+                instance_blk_off += INSTANCE_BLOCK_LEN;
             }
-            // offset to metric block
-            c.write_u64::<Endian>(metric_block_off)?;
-            // offset to instance block
-            c.write_u64::<Endian>(0)?;
-
-            // string block
-            let mut string_block_off = self.wi.string_sec_off
-                + self.wi.n_strings*STRING_BLOCK_LEN;
 
-            // short help
-            if m.shorthelp().len() > 0 {
-                c.set_position(shorthelp_off_off);
-                c.write_u64::<Endian>(string_block_off)?;
+            self.wi.instance_idx += im.metrics.len() as u64;
+            self.wi.indom_idx += 1;
 
-                c.set_position(string_block_off);
-                c.write_all(m.shorthelp().as_bytes())?;
-                c.write_all(&[0])?;
+            self.wi.indom_cache.insert(indom.id, instance_blk_offs);
+        }
 
-                self.wi.n_strings += 1;
-                string_block_off += STRING_BLOCK_LEN;
-            }
+        let instance_blk_offs = self.wi.indom_cache.get(&indom.id).unwrap().clone();
+        for ((_, metric), instance_blk_off) in im.metrics.iter_mut().zip(instance_blk_offs) {
+            self.__register_metric(metric, instance_blk_off)?;
+        }
 
-            // long help
-            if m.longhelp().len() > 0 {
-                c.set_position(longhelp_off_off);
-                c.write_u64::<Endian>(string_block_off)?;
+        Ok(self)
+    }
 
-                c.set_position(string_block_off);
-                c.write_all(m.longhelp().as_bytes())?;
-                c.write_all(&[0])?;
+    fn __register_metric<T: MetricType + Clone>(&mut self, m: &mut Metric<T>, instance_blk_off: u64) -> io::Result<&mut Client> {
```

Comment:
This writes an entire metrics block for each metric, which means things like semantics, units, type etc. are written individually each time which is incorrect behavior for instance metrics, as instance metrics share semantics, unit and other metadata and are written only once